### PR TITLE
fix(components): [select] label is not shown when value is a object

### LINKED ISSUE #7284: [Bug Report] [Component] [select] el-select组件绑定值为对象时，value-key和label属性只能为value和label值才生效，且回显失败
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.6`
- Element Plus Version: `2.1.10`
- Browser / OS: ` Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.88 Safari/537.36`
- Build Tool: `Webpack`

### Related Component

- `el-select`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuXG4gIDxlbC1zZWxlY3RcbiAgICB2LW1vZGVsPVwidmFsdWUyXCJcbiAgICBwbGFjZWhvbGRlcj1cIlNlbGVjdFwiXG4gICAgdmFsdWUta2V5PVwiY2lkXCJcbiAgICBAY2hhbmdlPVwiY2hhbmdlXCJcbiAgPlxuICAgIDxlbC1vcHRpb25cbiAgICAgIHYtZm9yPVwiaXRlbSBpbiBvcHRpb25zXCJcbiAgICAgIDpsYWJlbD1cIml0ZW0ubmFtZVwiXG4gICAgICA6dmFsdWU9XCJpdGVtXCJcbiAgICA+XG4gICAgPC9lbC1vcHRpb24+XG4gIDwvZWwtc2VsZWN0PlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdCBsYW5nPVwidHNcIiBzZXR1cD5cbmltcG9ydCB7IHNldHVwRWxlbWVudFBsdXMgfSBmcm9tICcuL2VsZW1lbnQtcGx1cy5qcydcbnNldHVwRWxlbWVudFBsdXMoKVxuXG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5cbmNvbnN0IHZhbHVlMSA9IHJlZihbJ09wdGlvbjEnXSlcbmNvbnN0IHZhbHVlMiA9IHJlZih7XG4gIH0pXG5jb25zdCBjaGFuZ2UgPSAoZGF0YSkgPT57XG4gIGNvbnNvbGUubG9nKGRhdGEpXG59XG5jb25zdCBvcHRpb25zID0gW1xuICB7XG4gICAgY2lkOiAnT3B0aW9uMScsXG4gICAgbmFtZTogJ09wdGlvbjEnLFxuICB9LFxuICB7XG4gICAgY2lkOiAnT3B0aW9uMicsXG4gICAgbmFtZTogJ09wdGlvbjInLFxuICB9LFxuICB7XG4gICAgY2lkOiAnT3B0aW9uMycsXG4gICAgbmFtZTogJ09wdGlvbjMnLFxuICB9LFxuICB7XG4gICAgY2lkOiAnT3B0aW9uNCcsXG4gICAgbmFtZTogJ09wdGlvbjQnLFxuICB9LFxuICB7XG4gICAgY2lkOiAnT3B0aW9uNScsXG4gICAgbmFtZTogJ09wdGlvbjUnLFxuICB9LFxuXVxuXG48L3NjcmlwdD5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcImltcG9ydHNcIjp7fX0ifQ==)

### Steps to reproduce

```vue
<template>

  <el-select
    v-model="value2"
    placeholder="Select"
    value-key="value"
    @change="change"
  >
    <el-option
      v-for="item in options"
      :label="item.label"
      :value="item"
    >
    </el-option>
  </el-select>
</template>

<script lang="ts" setup>
import { setupElementPlus } from './element-plus.js'
setupElementPlus()

import { ref } from 'vue'

const value1 = ref(['Option1'])
const value2 = ref({
  value: 'Option1'
})
const change = (data) =>{
  console.log(data)
}
const options = [
  {
    value: 'nane,cid',
    label: 'Option1',
  },
  {
    value: 'Option2',
    label: 'Option2',
  },
  {
    value: 'Option3',
    label: 'Option3',
  },
  {
    value: 'Option4',
    label: 'Option4',
  },
  {
    value: 'Option5',
    label: 'Option5',
  },
]

</script>

```

### What is Expected?

选择正确

### What is actually happening?

选择显示失效

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by exherb**
- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

cachedOption is a component instance

https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtc2VsZWN0IHYtbW9kZWw9XCJ2YWx1ZTFcIiBtdWx0aXBsZSB2YWx1ZS1rZXk9XCJ2YWx1ZVwiIHBsYWNlaG9sZGVyPVwiU2VsZWN0XCI+XG4gICAgPGVsLW9wdGlvblxuICAgICAgdi1mb3I9XCJpdGVtIGluIG9wdGlvbnNcIlxuICAgICAgOmtleT1cIml0ZW0udmFsdWVcIlxuICAgICAgOmxhYmVsPVwiaXRlbS50ZXN0XCJcbiAgICAgIDp2YWx1ZT1cIml0ZW1cIlxuICAgID5cbiAgICA8L2VsLW9wdGlvbj5cbiAgPC9lbC1zZWxlY3Q+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgc2V0dXBFbGVtZW50UGx1cyB9IGZyb20gJy4vZWxlbWVudC1wbHVzLmpzJ1xuc2V0dXBFbGVtZW50UGx1cygpXG5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgdmFsdWUxID0gcmVmKFtdKVxuY29uc3Qgb3B0aW9ucyA9IFtcbiAge1xuICAgIHZhbHVlOiAnT3B0aW9uMScsXG4gICAgdGVzdDogJ09wdGlvbjEnLFxuICB9LFxuICB7XG4gICAgdmFsdWU6ICdPcHRpb24yJyxcbiAgICB0ZXN0OiAnT3B0aW9uMicsXG4gIH0sXG4gIHtcbiAgICB2YWx1ZTogJ09wdGlvbjMnLFxuICAgIHRlc3Q6ICdPcHRpb24zJyxcbiAgfSxcbiAge1xuICAgIHZhbHVlOiAnT3B0aW9uNCcsXG4gICAgdGVzdDogJ09wdGlvbjQnLFxuICB9LFxuICB7XG4gICAgdmFsdWU6ICdPcHRpb241JyxcbiAgICB0ZXN0OiAnT3B0aW9uNScsXG4gIH0sXG5dXG5cbjwvc2NyaXB0PlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1wiaW1wb3J0c1wiOnt9fSJ9

fix #7284

### CODE REVIEW

**Comment by exherb** [comment]
I thinks `option value` / `v-model value` should keep same data struct. @btea

File: packages/components/select/src/useSelect.ts [hunk 1]
```diff
@@ -464,7 +464,7 @@ export const useSelect = (props, states: States, ctx) => {
     for (let i = states.cachedOptions.size - 1; i >= 0; i--) {
       const cachedOption = cachedOptionsArray.value[i]
       const isEqualValue = isObjectValue
-        ? get(cachedOption, props.valueKey) === get(value, props.valueKey)
+        ? get(cachedOption.value, props.valueKey) === get(value, props.valueKey)
```
Comment (btea) [inline packages/components/select/src/useSelect.ts:467]:
```suggestion
        ? get(
            isObject(cachedOption.value) ? cachedOption.value : cachedOption, 
            props.valueKey
          ) === get(value, props.valueKey)
```

**Comment by JeremyWuuuuu** [comment]
@exherb WDYT about the comment @btea left? I'd like to merge the ASAP so if you can update that it'd be great

**Comment by exherb** [comment]
> @exherb WDYT about the comment @btea left? I'd like to merge the ASAP so if you can update that it'd be great

Sorry, I will update now

File: packages/components/select/src/useSelect.ts (hunk 1 above)
Comment (exherb) [reply packages/components/select/src/useSelect.ts:467]: I don't understand why doing this check. It only bring complex and confuse.

The struct should be same.

Comment (btea) [reply packages/components/select/src/useSelect.ts:467]: I seem to have misunderstood your intention before, you mean that if `v-model` is object, `option.value` is also object...

Comment (btea) [reply packages/components/select/src/useSelect.ts:467]: If so, maybe it should be added to the documentation, what do you think? @JeremyWuuuuu

Comment (exherb) [reply packages/components/select/src/useSelect.ts:467]: I think this is the behavior before #6933. Is any reason (issue or something else) for #6933?

Comment (btea) [reply packages/components/select/src/useSelect.ts:467]: It has never been stated that `option.value` must be consistent with the `v-model` type, so if `v-model` is set to object, it needs to take effect.

Comment (btea) [reply packages/components/select/src/useSelect.ts:467]: You can take a look at this test
https://github.com/element-plus/element-plus/blob/99848c1623fe8a24c0b8408b29c5868e66c83e05/packages/components/select/__tests__/select.spec.ts#L364

**Comment by hitengwei** [comment]
image
image
Hi , May I ask if it is because of this problem that the display is not displayed here? It bothers me very much？？

File: packages/components/select/src/useSelect.ts (hunk 1 above)
Comment (exherb) [reply packages/components/select/src/useSelect.ts:467]: This test is added in #6933

Comment (btea) [reply packages/components/select/src/useSelect.ts:467]: I know, this test is the reason that created the pr #6933

Comment (exherb) [reply packages/components/select/src/useSelect.ts:467]: you can only choose one logic.

```
value-key = "c"
option = [{ value: "a", "label": "a"}]
v-model = {}

---
v-model = { "c" : "a" }

or 

v-model = "a"
```

```
value-key = "c"
option = [{ value: { "c": "a" }, "label": "a"}]
v-model = {}

---
v-model = undefined

or 

v-model = { "c": "a" }
```

Comment (exherb) [reply packages/components/select/src/useSelect.ts:467]: But if the team think compatible (little confusion) is a better solution, I'm ok with it.

Comment (exherb) [reply packages/components/select/src/useSelect.ts:467]: any update on this?

Comment (btea) [reply packages/components/select/src/useSelect.ts:467]: WDYT @JeremyWuuuuu

**Comment by btea** [comment]
@hitengwei Can you provide a link to this example?

File: packages/components/select/src/useSelect.ts (hunk 1 above)
Comment (exherb) [reply packages/components/select/src/useSelect.ts:467]: this is take too long. I think I will accept the suggestion code directly tomorrow.

**Comment by exherb** [comment]
@btea @JeremyWuuuuu please review, thanks ~

**Comment by btea** [comment]
@exherb The test seems to fave failed, please run `pnpm lint`.  

And Can you add a test about this situation? #7141 

Thank you.

**Comment by exherb** [comment]
> @exherb The test seems to fave failed, please run `pnpm lint`.
> 
> And Can you add a test about this situation? #7141
> 
> Thank you.

OK, I will take a look

**Comment by L1yp** [comment]
2.1.10 版本该问题尚未修复，
对此功能是刚需，用户选择 部门选择 很有用，双击打开对话框选择，传回object可以让options不存在也能正确显示

**Comment by jyp-qiyu** [comment]
该问题还未修复，当v-model绑定值未对象是，el-option的label值必须是label，无法使用自定义的对象名，否则就不显示选中的label

**Comment by Lovercz** [comment]
刚好我们项目在测试，那感觉酸爽~

**Comment by m3xTa1nes** [comment]
Same problem! Not fixed in 2.1.10

**Comment by btea** [comment]
@exherb Hi, do you have time to fix this? If not, we will close the pr and recreate a new pr to fix it.

**Comment by exherb** [comment]
> 

OK...

**Comment by exherb** [comment]
> @exherb Hi, do you have time to fix this? If not, we will close the pr and recreate a new pr to fix it.

Hi, can I rollback to my init commit? Since https://github.com/element-plus/element-plus/pull/7288 can be approve.

**Comment by btea** [comment]
> Hi, can I rollback to my init commit? Since #7288 can be approve.

Hi, thank you very much for your contribution. That pr has already been approved, so you don't need to roll back the code.

However, there seems to be a lack of relevant tests for the usage scenarios of these issues. Would you like to add them?

**Comment by exherb** [comment]
> > Hi, can I rollback to my init commit? Since #7288 can be approve.
> 
> Hi, thank you very much for your contribution. That pr has already been approved, so you don't need to roll back the code.
> 
> However, there seems to be a lack of relevant tests for the usage scenarios of these issues. Would you like to add them?

I rollback to my init commit and added test.

**Comment by exherb** [comment]
@btea the unit test has passed, got time to review?

**Review by btea — APPROVED** [review]

File: packages/components/select/__tests__/select.test.ts [hunk 2]
```diff
@@ -393,6 +397,39 @@ describe('Select', () => {
...
+          {
+            id: 1,
+            name: '黄金糕',
+          },
+          {
+            id: 2,
+            name: '双皮奶',
+          },
+        ],
+        value: {
```
Comment (btea) [inline packages/components/select/__tests__/select.test.ts:423]: the `value` maybe should not a object
```suggestion
        value: 2
```

Comment (exherb) [reply packages/components/select/__tests__/select.test.ts:423]: you are right.

Comment (exherb) [reply packages/components/select/__tests__/select.test.ts:454]: the default value-key is `value`, I added another test

**Comment by exherb** [comment]
are the team going to release a version recently?

**Comment by jyp-qiyu** [comment]
2.1.11版本还是没有修复这个问题，请问下个版本会进行修复吗

**Comment by exherb** [comment]
> 2.1.11版本还是没有修复这个问题，请问下个版本会进行修复吗

如果合并了，下个版本可能就会被修复

**Comment by JeremyWuuuuu** [comment]
Hi all, the next version will include this PR inside, a little bit busy making virtual table right now. Sorry for the delay.

**Pull request merged**